R/formula.R

```diff
@@ -0,0 +1,407 @@
+count_terms_in_subformula <- function(subformula) {
+  if (!inherits(subformula, "formula"))
+    subformula <- as.formula(paste0("~ ", subformula))
+  tt <- extract_terms_response(subformula)
+  ind_interaction_terms <- length(tt$individual_terms) + length(tt$interaction_terms)
+  group_terms <- sum(unlist(lapply(tt$group_terms, count_terms_in_group_term)))
+  ind_interaction_terms + group_terms
+}
+
+#' Utility to count the number of terms in a given group term.
+#' @param term A group term as a string.
+#' @return the number of terms in the group term.
+count_terms_in_group_term <- function(term) {
+  term <- gsub("[\\(\\)]", "", flatten_group_terms(term))
+
+  chunks <- strsplit(term,  "[ ]*\\|([^\\|]*\\||)[ ]*")[[1]]
+  lhs <- as.formula(paste0("~", chunks[1]))
+  tt <- extact_terms_response(lhs)
+  variables <- c(tt$individual_terms, tt$interaction_terms)
+
+  ## split possible interaction terms
+  int_t <- grepl(":", variables)
+  int_v <- variables[int_t]
+  lin_v <- variables[!int_t]
+
+  if ("0" %in% variables) {
+    group_intercept <- FALSE
+    variables <- setdiff(variables, "0")
+  } else {
+    group_intercept <- TRUE
+    if (!("1" %in% variables))
+      variables <- c(variables, "1")
+  }
+
+  length(variables)
+}
+
+#' Given a list of formulas, sort them by number of terms in them.
+#' @param submodels A list of models' formulas.
+#' @return a sorted list of submodels by included terms.
+sort_submodels_by_size <- function(submodels) {
+  size <- lapply(submodels, count_terms_in_subformula)
+  df <- as.data.frame(cbind(submodels=submodels, size=unlist(size)))
+  df$size <- as.numeric(df$size)
+  ordered <- df[order(df$size), ]
+
+  groups <- list()
+  for (size in unique(ordered$size))
+    groups[[size]] <- as.character(ordered$submodels[ordered$size == size])
+
+  ord_list <- groups
+  ## remove NA inside submodels
+  ord_list_nona <- lapply(ord_list, function(l) l[!is.na(l)])
+  ## remove NA at the submodels level
+  ord_list_nona[!is.na(ord_list_nona)]
+}
+
+#' Given a refmodel structure, count the number of variables included.
+#' @param formula The reference model's formula.
+#' @param list_of_terms Subset of terms from formula.
+#' @return number of variables in list_of_terms.
+count_variables_chosen <- function(formula, list_of_terms) {
+  if (length(list_of_terms) == 0)
+    return(0)
+  formula <- subset_formula(formula, unique(unlist(list_of_terms)))
+  count_terms_in_subformula(flatten_formula(formula))
+}
+
+#' Utility that checks if the next submodel is redundant with the current one.
+#' @param formula The reference models' formula.
+#' @param current A list of terms included in the current submodel.
+#' @param new The new term to add to the submodel.
+#' @return TRUE if the new term results in a redundant model, FALSE otherwise.
+is_next_submodel_redundant <- function(formula, current, new) {
+  old_submodel <- current
+  new_submodel <- c(current, new)
+  if (count_variables_chosen(formula, new_submodel) >
```

Comment:
I know we discussed this, but how can we be sure some model is redundant just because it has less terms? After all, it could contain *different* terms?